src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs

```diff
@@ -318,7 +318,27 @@ public JoinableTask RunAsync(Func<Task> asyncMethod)
     /// </remarks>
     public JoinableTask RunAsync(Func<Task> asyncMethod, JoinableTaskCreationOptions creationOptions)
     {
-        return this.RunAsync(asyncMethod, synchronouslyBlocking: false, creationOptions: creationOptions);
+        return this.RunAsync(asyncMethod, synchronouslyBlocking: false, parentToken: null, creationOptions: creationOptions);
+    }
+
+    /// <summary>
+    /// Invokes an async delegate on the caller's thread, and yields back to the caller when the async method yields.
+    /// The async delegate is invoked in such a way as to mitigate deadlocks in the event that the async method
+    /// requires the main thread while the main thread is blocked waiting for the async method's completion.
+    /// </summary>
+    /// <param name="asyncMethod">The method that, when executed, will begin the async operation.</param>
+    /// <param name="parentToken">An optional token that identifies one or more <see cref="JoinableTask"/> instances, typically in other processes, that serve as 'parents' to this one.</param>
+    /// <param name="creationOptions">The <see cref="JoinableTaskCreationOptions"/> used to customize the task's behavior.</param>
+    /// <returns>An object that tracks the completion of the async operation, and allows for later synchronous blocking of the main thread for completion if necessary.</returns>
+    /// <remarks>
+    /// <para>Exceptions thrown by the delegate are captured by the returned <see cref="JoinableTask" />.</para>
+    /// <para>When the delegate resumes from a yielding await, the default behavior is to resume in its original context
+    /// as an ordinary async method execution would. For example, if the caller was on the main thread, execution
+    /// resumes after an await on the main thread; but if it started on a threadpool thread it resumes on a threadpool thread.</para>
+    /// </remarks>
+    public JoinableTask RunAsync(Func<Task> asyncMethod, string? parentToken, JoinableTaskCreationOptions creationOptions)
```

Comment:
Add a `RunAsync<T>` overload to simplify the (only) use case we have, which is in StreamJsonRpc.